modules/network/src/main/java/org/apache/ignite/internal/network/serialization/marshal/BuiltInContainerMarshallers.java

```diff
@@ -0,0 +1,198 @@
+        output.writeUTF(array.getClass().getComponentType().getName());
+        return writeCollection(Arrays.asList(array), arrayDescriptor, output);
+    }
+
+    <T> T[] readGenericRefArray(DataInput input, ValueReader<T> elementReader)
+            throws IOException, UnmarshalException {
+        return BuiltInMarshalling.readGenericRefArray(input, elementReader);
+    }
+
+    List<ClassDescriptor> writeBuiltInCollection(Collection<?> object, ClassDescriptor descriptor, DataOutput output)
+            throws IOException, MarshalException {
+        if (supportsAsMutableBuiltInCollection(descriptor)) {
+            return writeCollection(object, descriptor, output);
+        } else if (descriptor.isSingletonList()) {
+            return writeSingletonList(object, descriptor, output);
+        } else {
+            throw new IllegalStateException("Marshalling of " + descriptor.clazz() + " is not supported, but it's marked as a built-in");
+        }
+    }
+
+    /**
+     * Returns {@code true} if the given descriptor is supported as a built-in mutable collection. Such types
+     * are eligible for a generic unmarshal algorithm: read length, create an empty collection, then read N elements
+     * and add each of them into the collection.
+     *
+     * @param descriptor the descriptor to check
+     * @return {@code true} if the given descriptor is supported as a built-in mutable collection
+     */
+    private boolean supportsAsMutableBuiltInCollection(ClassDescriptor descriptor) {
+        return mutableBuiltInCollectionFactories.containsKey(descriptor.clazz());
+    }
+
+    private List<ClassDescriptor> writeCollection(Collection<?> collection, ClassDescriptor collectionDescriptor, DataOutput output)
+            throws IOException, MarshalException {
+        Set<ClassDescriptor> usedDescriptors = new HashSet<>();
+        usedDescriptors.add(collectionDescriptor);
+
+        BuiltInMarshalling.writeCollection(collection, output, writerAddingUsedDescriptor(usedDescriptors));
+
+        return List.copyOf(usedDescriptors);
+    }
+
+    private <T> ValueWriter<T> writerAddingUsedDescriptor(Set<ClassDescriptor> usedDescriptors) {
+        return (elem, out) -> {
+            List<ClassDescriptor> elementDescriptors = trackingMarshaller.marshal(elem, out);
+            usedDescriptors.addAll(elementDescriptors);
+        };
+    }
+
+    private List<ClassDescriptor> writeSingletonList(Collection<?> collection, ClassDescriptor listDescriptor, DataOutput output)
+            throws MarshalException, IOException {
+        assert collection.size() == 1;
+
+        Object element = collection.iterator().next();
+
+        Set<ClassDescriptor> usedDescriptors = new HashSet<>();
+        usedDescriptors.add(listDescriptor);
+
+        List<ClassDescriptor> descriptorsFromElement = trackingMarshaller.marshal(element, output);
+        usedDescriptors.addAll(descriptorsFromElement);
+
+        return List.copyOf(usedDescriptors);
```

Comment:
Why not simply return `usedDescriptors` set?

Reply:
Changed from `List` to `Set`, so no copying here is needed anymore